lib/stacks/iam.ts

```diff
@@ -396,12 +405,65 @@ export class IAMStack extends Stack {
             return statements;
         };
 
+        /*
+         * WARNING: Changing this method will cause any policy statement created by this to be regenerated. This will cause
+         * any changes to this policy (like dynamic policy updates for app configuration changes) to be lost until the app
+         * configuration is updated and it updates this policy with the expected values.
+         */
+        const instanceConstraintPolicyStatement = (partition: string, region: string, actionResourcePair: {[key: string]: string}) => {
+            const [actions, resources] = Object.entries(actionResourcePair).reduce(([actionAccumulator, resourcesAccumulator], [action, resource]) => {
+                return [[...actionAccumulator, `sagemaker:${action}`], [...resourcesAccumulator, `arn:${partition}:sagemaker:${region}:${this.account}:${resource}/*`]];
+            }, [[] as string[], [] as string[]]);
+
+            return [
+                new PolicyStatement({
+                    effect: Effect.ALLOW,
+                    actions,
+                    resources,
+                    conditions: {
+                        'ForAnyValue:StringEquals': {
+                            'sagemaker:InstanceTypes': [],
+                        },
+                    },
+                }),
+            ];
+        };
+
         const notebookPolicy = new ManagedPolicy(this, 'mlspace-notebook-policy', {
             statements: notebookPolicyStatements(this.partition, Aws.REGION),
             description: 'Enables general MLSpace actions in notebooks and across the entire application.'
         });
 
-        // If roles are manually created use the existing role
+        if (props.mlspaceConfig.ENDPOINT_CONFIG_INSTANCE_CONSTRAINT_POLICY_ARN) {
```

Comment:
This if-else conditional and the other should be within a conditional checking for `MANAGE_IAM_ROLES`. These policies can't / shouldn't be created if dynamic roles are not being used.

Reply:
Good callout although I'm not 100% sure it needs to be checking for `MANAGE_IAM_ROLES`. My understanding was `MANAGE_IAM_ROLES` was for dynamic user roles not just anywhere we're updating roles dynamically.

I'm going to move these two conditions inside the else condition below for when the notebook role isn't explicitly defined. This will then follow the same pattern as was already being done for the notebook role.

Reply:
After speaking offline I see why this is necessary and will make these changes.